### PR TITLE
test: Code Coverage: Add Edit Mileage Lint #2

### PR DESCRIPTION
**Pull request by jayfyle**
### <samp>🤖 Generated by Copilot at a9c41fd</samp>

This pull request adds support for displaying file thumbnails and location data in the expense form, improves the type safety and clarity of various interfaces and properties related to transactions, cost centers, mileage rates, and expense fields, and enhances the code quality of some services by adding return types and removing unused imports.

### <samp>🤖 Generated by Copilot at a9c41fd</samp>

> _Sing, O Muse, of the valiant code review_
> _That refined the types and interfaces_
> _Of the expense form and the transactions_
> _With skill and care worthy of Hephaestus._

### <samp>🤖 Generated by Copilot at a9c41fd</samp>

*  Define a new interface `CostCenterOptions` for displaying cost center options in the expense form ([link](https://github.com/fylein/fyle-mobile-app/pull/2230/files?diff=unified&w=0#diff-27ef7dd1e2ecbd89a5c72dea4f238188eeea1e67398a3351d73c9f3cb80f4c36R1-R6))
*  Import the `CostCenter` and `Location` interfaces from their respective files and use them to specify the types of the `value` property in `CostCenterOptions` and the `locations` property in `Transaction` ([link](https://github.com/fylein/fyle-mobile-app/pull/2230/files?diff=unified&w=0#diff-27ef7dd1e2ecbd89a5c72dea4f238188eeea1e67398a3351d73c9f3cb80f4c36R1-R6), [link](https://github.com/fylein/fyle-mobile-app/pull/2230/files?diff=unified&w=0#diff-db96e89a4671a6577872d9c95a3722abc4d79e6b3f8443ec4f74cedb81915dbaR2), [link](https://github.com/fylein/fyle-mobile-app/pull/2230/files?diff=unified&w=0#diff-db96e89a4671a6577872d9c95a3722abc4d79e6b3f8443ec4f74cedb81915dbaL36-R37))
*  Modify the types of the `value` property in `TxnCustomProperties` and the `options` property in `ExpenseField` to be more specific and handle different kinds of values and options for the custom fields and expense fields ([link](https://github.com/fylein/fyle-mobile-app/pull/2230/files?diff=unified&w=0#diff-ee5bdbbd7c8ccd3a96a4cd0e7ef7dc894df760a6f150ae516a94c6457bcc8bf9L9-R9), [link](https://github.com/fylein/fyle-mobile-app/pull/2230/files?diff=unified&w=0#diff-847e5f64ad45ea9d6680562186e2abdf75829fd94ece9751239af32a5745dda7L14-R14))
*  Import the `FileObject` interface from its file and use it to specify the type of the `dataUrls` property in `UnflattenedTransaction`, which stores the data urls of the files attached to the transactions ([link](https://github.com/fylein/fyle-mobile-app/pull/2230/files?diff=unified&w=0#diff-173c888e51f0c760dd978d56b4f3848c035ae1ee0fca9eb9b91726349040bf1bR1), [link](https://github.com/fylein/fyle-mobile-app/pull/2230/files?diff=unified&w=0#diff-173c888e51f0c760dd978d56b4f3848c035ae1ee0fca9eb9b91726349040bf1bR26))
*  Modify the `MileageSettings` interface to have a more specific type for the `mileage_rate_labels` property, which stores the custom labels for the mileage rates ([link](https://github.com/fylein/fyle-mobile-app/pull/2230/files?diff=unified&w=0#diff-62bc2c417b18ca7598db8ba60eb3da941fb059aca0d64a4c485f68cd5469a229L21-R21))
*  Modify the `getEtxnSelectedPaymentMode` and `checkIfEtxnHasSamePaymentMode` methods of the `AccountsService` class to accept partial `UnflattenedTransaction`s as parameters, to allow them to work with transactions that have not been fully populated yet ([link](https://github.com/fylein/fyle-mobile-app/pull/2230/files?diff=unified&w=0#diff-acb4f2d0105052014c5b8746f35020b2ecf00198ec7b1d278c6f172704ac62e1L82-R82), [link](https://github.com/fylein/fyle-mobile-app/pull/2230/files?diff=unified&w=0#diff-acb4f2d0105052014c5b8746f35020b2ecf00198ec7b1d278c6f172704ac62e1L186-R186))
*  Add return types to the `getReadableRate` method of the `MileageRatesService` class and the `getFilteredPendingReports` method of the `ReportService` class, to make the method signatures more explicit ([link](https://github.com/fylein/fyle-mobile-app/pull/2230/files?diff=unified&w=0#diff-3c789edd12fc41da591bd9c248fc12c338269f4526c520850f6146a1f33e8813L80-R80), [link](https://github.com/fylein/fyle-mobile-app/pull/2230/files?diff=unified&w=0#diff-12dc98777c493c39a3e7aec29b439cc7863ddfb002bf1e52052c8aee485c63c9L569-R569))
*  Remove the unused import of `tap` from the `mileage-rates.service.ts` file, to improve the code quality and avoid unnecessary imports ([link](https://github.com/fylein/fyle-mobile-app/pull/2230/files?diff=unified&w=0#diff-3c789edd12fc41da591bd9c248fc12c338269f4526c520850f6146a1f33e8813L8-R8))

## Clickup
https://app.clickup.com/t/85ztk92ck

## Code Coverage
Please add code coverage here

## UI Preview
Please add screenshots for UI changes

### CODE REVIEW

**Review by Dimple16 — CHANGES_REQUESTED** [review]
I’ll review once the lint, and unit test issues are resolved, please fix and reassign

File: src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts [hunk 1]
```diff
@@ -1876,7 +1973,7 @@ export class AddEditMileagePage implements OnInit {
     );
   }
 
-  async continueWithCriticalPolicyViolation(criticalPolicyViolations: string[]) {
+  async continueWithCriticalPolicyViolation(criticalPolicyViolations: string[]): Promise<unknown> {
```
Comment (Dimple16) [inline src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts:1976]: please don't use `unknown`

File: src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts [hunk 2]
```diff
@@ -1906,11 +2008,17 @@ export class AddEditMileagePage implements OnInit {
 
     await currencyModal.present();
 
-    const { data } = await currencyModal.onWillDismiss();
+    const { data } = (await currencyModal.onWillDismiss()) as {
+      data: unknown;
```
Comment (Dimple16) [inline src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts:2012]: same here

File: src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts [hunk 3]
```diff
@@ -1889,11 +1986,16 @@ export class AddEditMileagePage implements OnInit {
...
+    };
     return !!data;
   }
 
-  async continueWithPolicyViolations(policyViolations: string[], policyAction: FinalExpensePolicyState) {
+  async continueWithPolicyViolations(
+    policyViolations: string[],
+    policyAction: FinalExpensePolicyState
+  ): Promise<unknown> {
```
Comment (Dimple16) [inline src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts:1998]: same here

File: src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts [hunk 4]
```diff
@@ -2074,44 +2188,52 @@ export class AddEditMileagePage implements OnInit {
...
+              switchMap((continueWithTransaction) => {
+                if (continueWithTransaction) {
+                  return from(this.loaderService.showLoader()).pipe(switchMap(() => of({ etxn: err.etxn })));
+                } else {
+                  return throwError('unhandledError');
+                }
+              })
+            );
+          } else if (err.type === 'policyViolations') {
+            return from(this.continueWithPolicyViolations(err.policyViolations, err.policyAction)).pipe(
```
Comment (Dimple16) [inline src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts:2215]: this entrire block can be moved to a separate method if you want to consider refactoring while you're in this area

File: src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts [hunk 5]
```diff
@@ -2299,42 +2423,50 @@ export class AddEditMileagePage implements OnInit {
...
+          etxn?: UnflattenedTransaction;
+          policyAction?: FinalExpensePolicyState;
+        }) => {
+          if (err.status === 500) {
+            return this.generateEtxnFromFg(this.etxn$, customFields$, calculatedDistance$).pipe(
+              map((etxn) => ({ etxn }))
+            );
+          }
+          if (err.type === 'criticalPolicyViolations') {
+            return from(this.continueWithCriticalPolicyViolation(err.policyViolations)).pipe(
```
Comment (Dimple16) [inline src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts:2440]: same comments,
we can have 2 methods for the below, `handleCriticalPolicyViolations()` and `handlePolicyViolations()`

File: src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts [hunk 6]
```diff
@@ -2466,7 +2611,11 @@ export class AddEditMileagePage implements OnInit {
 
     await modal.present();
 
-    const { data } = await modal.onDidDismiss();
+    const { data } = (await modal.onDidDismiss()) as {
+      data: {
+        updated: unknown;
```
Comment (Dimple16) [inline src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts:2616]: avoid `unknown`

**Review by Dimple16 — APPROVED** [review]
added couple of comments, fix and merge

File: src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts (hunk 4 above)
Comment (jayfyle) [reply src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts:2215]: Yes, will add this in a separate PR after this

File: src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts (hunk 5 above)
Comment (jayfyle) [reply src/app/fyle/add-edit-mileage/add-edit-mileage.page.ts:2440]: Will add this in the next PR, will refactor it similar to add edit expense

**Review by jayfyle — COMMENTED** [review]
Some places have `unknown`  as type, because the type will change after refactoring for the methods. Since I will be doing the refactoring in the upcoming PR, will fix the type issue in the same PR.

**Review by Dimple16 — APPROVED** [review]

**Pull request merged**